tsdb/index/postings.go

```diff
@@ -321,11 +306,14 @@ func (p *MemPostings) Delete(deleted map[storage.SeriesRef]struct{}) {
 				}
 			}
 			if !found {
-				p.mtx.Unlock()
 				continue
 			}
+
 			repl := make([]storage.SeriesRef, 0, len(p.m[n][l]))
 
+			// Only take the write lock when there's actually something to write.
+			p.mtx.RUnlock()
```

Comment:
Unlocking here means another goroutine can jump in and alter `p.m`, which is unsafe because we are part-way through `for n := range p.m` on line 297.
I suspect this is why the original code copied the keys and values.

Reply:
Thanks, good point. Reverted copying of keys & values to their slices in https://github.com/prometheus/prometheus/pull/13286/commits/b68c9c7a734996642ed79edc03dd666285492fd1

Also added proper estimation of slice sizes to avoid re-allocations.